packages/cli/src/rpc/explain/navie/history.ts

```diff
@@ -0,0 +1,345 @@
+    try {
+      await readlink(path);
+    } catch (e) {
+      const err = e as Error & { code?: string };
+      if (err.code === 'ENOENT') {
+        await symlink(target, path);
+      } else if (err.code === 'EEXIST') {
+        // Symlink already exists, do nothing.
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  private async ensureThreadDir(threadId: string): Promise<string> {
+    const threadDir = join(this.directory, 'threads', threadId);
+    await mkdir(threadDir, { recursive: true });
+    const projectDirectoriesFile = join(threadDir, 'projectDirectories.txt');
+    if (!(await exists(projectDirectoriesFile))) {
+      const projectDirectories = configuration().projectDirectories;
+      await writeFile(projectDirectoriesFile, projectDirectories.join('\n'));
+    }
+    return threadDir;
+  }
+
+  static threadAccessError(threadId: string, action: string, e: any): ThreadAccessError {
+    const err = e as Error & { code?: string };
+    if (err.code === 'ENOENT') warn(`Thread ${threadId} not found`);
+
+    return new ThreadAccessError(threadId, action, e instanceof Error ? e : undefined);
+  }
+
+  // In the old-style history format, threads are stored in files named by their thread id
+  // and timestamp. In the new-style history format, threads are stored in files named by
+  // their thread id, and the thread file is symlinked to a file named for the date of the thread.
+  static async migrate(
+    oldDirectory: string,
+    history: History,
+    options: { cleanup: boolean } = { cleanup: true }
+  ): Promise<void> {
+    // List the contents of the old directory. Each one is a threadId directory, containing
+    // timestamped messages.
+    const threadIds = (await readdir(oldDirectory)).filter(
+      // Match UUIDs like 5278527e-c4ed-4e45-9fb6-372ed6a036f6 in the thread dir
+      // Being careful not to touch anything else, like the dates and threads directories that are
+      // created in the new history format.
+      (dir) => dir.match(THREAD_ID_REGEX)
+    );
+
+    const threadFileAsTimestamp = (threadFile: string) => parseInt(threadFile.split('.')[0]);
+
+    for (const threadId of threadIds) {
+      warn(`[history] Migrating thread ${threadId} from old history format`);
+
+      const oldThreadDir = join(oldDirectory, threadId);
+      const threadFiles = await readdir(oldThreadDir);
+      threadFiles.sort((a, b) => threadFileAsTimestamp(a) - threadFileAsTimestamp(b));
+
+      let lastUserMessageId: string | undefined;
+      for (const threadFile of threadFiles) {
+        const timestamp = threadFileAsTimestamp(threadFile);
+        // MessageIds are not available, so fake them with timestamps.
+        const messageId = timestamp.toString();
+        const messageFile = join(oldThreadDir, threadFile);
+        const messageStr = await readFile(messageFile, 'utf-8');
+        const message = JSON.parse(messageStr) as Message;
```

Comment:
One final issue here.

I'm seeing cases where this raises an exception. I believe some of these files could have been empty, which would result in `SyntaxError: Unexpected end of JSON input`.

I've lost the files now so I can't recreate this 100%.